content/posts/2024-04-02-cn9130-temp-regulation.md

```diff
@@ -0,0 +1,251 @@
+---
+title:  "Marvell CN9130 Temperature Regulation"
+author: Tobias Waldekranz
+date:   2024-04-02 18:54:00 +0100
+tags:
+ - linux
+ - cn9130
+ - thermal
+---
+
+Seeing as we are deploying the [CN9130][1] SoC from Marvell in
+multiple products targeting industrial applications, we wanted to
+explore how it reacts when exposed to high ambient temperatures.
+
+Software wise we are running a more or less unmodified Linux 6.6.22
+kernel, using the standard `performance` governor to manage CPU
+frequency scaling.
+
+_TL;DR: The system will will start throttling down the frequency of a
+core, from the maximum of 1.6 GHz, to 800 MHz when it reaches 85°C. If
+the core temperature exceeds 95°C, then the frequency is brought down
+to 400 MHz. If either die on the chip reports a temperature in excess
+of 100°C, an emergency shutdown is triggered to protect the hardware
+from permanent damage._
+
+
+## Passive Temperature Regulation
+
+Looking at the relevant parts of the SoC's [device tree][2], we find
+the following section:
+
+```c
+...
+	thermal-zones {
+		...
+		ap_thermal_cpu0: ap-cpu0-thermal {
+			polling-delay-passive = <1000>;
+			polling-delay = <1000>;
+
+			thermal-sensors = <&ap_thermal 1>;
+
+			trips {
+				cpu0_hot: cpu0-hot {
+					temperature = <85000>;
+					hysteresis = <2000>;
+					type = "passive";
+				};
+				cpu0_emerg: cpu0-emerg {
+					temperature = <95000>;
+					hysteresis = <2000>;
+					type = "passive";
+				};
+			};
+
+			cooling-maps {
+				map0_hot: map0-hot {
+					trip = <&cpu0_hot>;
+					cooling-device = <&cpu0 1 2>,
+						<&cpu1 1 2>;
+				};
+				map0_emerg: map0-ermerg {
+					trip = <&cpu0_emerg>;
+					cooling-device = <&cpu0 3 3>,
+						<&cpu1 3 3>;
+				};
+			};
+		};
+		...
+	};
+...
+```
+
+This defines the policy of how the kernel's [thermal][3] subsystem is
+allowed to throttle the CPU frequency of a core in response to high
+temperature environments. First, two trip points are defined:
+
+- `cpu0-hot` at 85°C
+- `cpu0-emerg` at 95°C
+
+These are serve as triggers for enforcing different CPU frequency
+limitations via the set of defined `cooling-maps`. Deciphering this,
+we see that:
+
+- When `cpu0-hot` trips, i.e. core 0's temperature reaches 85°C, then
+  we enable two `cooling-device`s (`cpu0` and `cpu1`). At this point
```

Comment:
`cooling devices`

Reply:
I was going for a plural of the property name in the device tree, hence the trailing "s" outside of the verbatim portion. But I think it just complicates things. Thanks!